flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/FlinkRuleSets.scala

```diff
@@ -29,50 +29,45 @@ object FlinkRuleSets {
     */
   val DATASET_OPT_RULES: RuleSet = RuleSets.ofList(
 
-    // filter rules
+    // push a filter into a join
     FilterJoinRule.FILTER_ON_JOIN,
+    // push filter into the children of a join
     FilterJoinRule.JOIN,
-    FilterMergeRule.INSTANCE,
-    FilterAggregateTransposeRule.INSTANCE,
+    // push filter through an aggregation
+    FlinkFilterAggregateTransposeRule.INSTANCE,
 
-    // push and merge projection rules
+    // aggregation and projection rules
     AggregateProjectMergeRule.INSTANCE,
-    ProjectMergeRule.INSTANCE,
+    AggregateProjectPullUpConstantsRule.INSTANCE,
+    // push a projection past a filter or vice versa
     ProjectFilterTransposeRule.INSTANCE,
     FilterProjectTransposeRule.INSTANCE,
-    AggregateProjectPullUpConstantsRule.INSTANCE,
-    JoinPushExpressionsRule.INSTANCE,
+    // push a projection to the children of a join
     ProjectJoinTransposeRule.INSTANCE,
+    // remove identity project
     ProjectRemoveRule.INSTANCE,
+    // reorder sort and projection
     SortProjectTransposeRule.INSTANCE,
     ProjectSortTransposeRule.INSTANCE,
 
-    // merge and push unions rules
-    // TODO: Add a rule to enforce binary unions
+    // join rules
+    JoinPushExpressionsRule.INSTANCE,
+
+    // remove union with only a single child
     UnionEliminatorRule.INSTANCE,
-    FlinkJoinUnionTransposeRule.LEFT_UNION,
-    FlinkJoinUnionTransposeRule.RIGHT_UNION,
```

Comment:
Alright. Then `FlinkJoinUnionTransposeRule` can be removed since it's not used. I can do that before merging.